### DISCUSSION TITLE
How to define constraints in miUML

### DISCUSSION THREAD

**Discussion by msche** [modelint/shlaer-mellor-metamodel#53] [Q&A] [answered]
I just started looking into executable UML and miUML and was wondering how constraints are defined within the metamodel. I couldn find it in the documentation and was kind of expecting that a Identifier would have (implicitly) a constraint that all attributes defined in the indentifier must be unique for the instances of the type in which it is defined

**Comment by modelint** [comment, marked as answer]
This is another great question. As you point out, there are certain implicit constraints that are built in to Shlaer-Mellor. The two key constraints that come to mind are identifiers and referential attributes. OCL definitions are available in [\[MB\]](https://github.com/modelint/shlaer-mellor-metamodel/wiki/Resources#mb-executable-uml-a-foundation-for-model-driven-architecture-stephen-mellor-mark-balcer-addison-wesley-2002), though I prefer a more relational approach as described in [\[MATHDP\]](https://github.com/modelint/shlaer-mellor-metamodel/wiki/Resources#mathdp-applied-mathematics-for-database-professionals-lex-de-haan-toon-koppelaars). Chapter 6 in particular introduces table, tuple and database predicates which, for us, translate into class, instance and domain predicates. This approach gives you a lot of power without the OCL bulk. More to the point, Shlaer-Mellor class and action models are an application of relational theory so I see a much tighter fit.

Strategic definition of identifiers and referential attributes eliminates the need for many explicit constraints. But there are certainly cases where you need to be spell things out. My best guess is that constraints merit their own domain. I am pondering something along these lines.

**Reply by msche** [reply]
Thanks for the answer. I'm going slowly through the book regarding the relational theory. When I'm done with that I will have a look again at the model with that knowledge in the back of my mind.
